libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
+  #      information is lost and must be logged here instead of being
+  #      available to the caller - further refactoring should propagate errors
+  #      to the caller instead
+  proc impl(
+      self: TcpTransport
+  ): Future[Connection] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if not self.running:
+      raise newTransportClosedError()
 
     if self.acceptFuts.len <= 0:
-      return
+      self.acceptFuts = self.servers.mapIt(it.accept())
 
     let
-      finished = await one(self.acceptFuts)
+      finished =
+        try:
+          await one(self.acceptFuts)
+        except ValueError:
+          raise (ref TcpTransportError)(msg: "No listeners configured")
+
       index = self.acceptFuts.find(finished)
+      transp =
+        try:
+          await finished
+        except TransportTooManyError as exc:
+          debug "Too many files opened", exc = exc.msg
+          return nil
+        except TransportAbortedError as exc:
+          debug "Connection aborted", exc = exc.msg
+          return nil
+        except TransportUseClosedError as exc:
+          raise newTransportClosedError(exc)
+        except TransportOsError as exc:
+          raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+        except common.TransportError as exc: # Needed for chronos 4.0.0 support
+          raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+        except CancelledError as exc:
+          raise exc
+
+    if not self.running: # Stopped while waiting
+      await transp.closeWait()
+      raise newTransportClosedError()
 
     self.acceptFuts[index] = self.servers[index].accept()
 
-    let transp = await finished
-    try:
-      let observedAddr = MultiAddress.init(transp.remoteAddress).tryGet()
-      return await self.connHandler(transp, Opt.some(observedAddr), Direction.In)
-    except CancelledError as exc:
-      debug "CancelledError", exc = exc.msg
-      transp.close()
-      raise exc
-    except CatchableError as exc:
-      debug "Failed to handle connection", exc = exc.msg
-      transp.close()
-  except TransportTooManyError as exc:
-    debug "Too many files opened", exc = exc.msg
-  except TransportAbortedError as exc:
-    debug "Connection aborted", exc = exc.msg
-  except TransportUseClosedError as exc:
-    debug "Server was closed", exc = exc.msg
-    raise newTransportClosedError(exc)
-  except CancelledError as exc:
-    raise exc
-  except TransportOsError as exc:
-    info "OS Error", exc = exc.msg
-    raise exc
-  except CatchableError as exc:
-    info "Unexpected error accepting connection", exc = exc.msg
-    raise exc
+    let remote =
+      try:
+        transp.remoteAddress
+      except TransportOsError as exc:
+        # The connection had errors / was closed before `await` returned control
+        await transp.closeWait()
+        debug "Cannot read remote address", exc = exc.msg
+        return nil
```

Comment:
why do we start returning `nil` here?

Reply:
it was always returning nil, by not returning anything (implicitly)